### ISSUE TITLE
Need explanation on Lab6 Level 3 getGrade() method

### ISSUE THREAD

**Issue by Earthstomp** [nus-cs2030/2122-s2#134]
Hi I have managed to pass level 3 already. However I am a bit confused on the individual return types within the maps and the final one using orElse(). Attached my method below:

String getGrade(String student, String module, String assessment) {
String noRecord = String.format("No such record: %s %s %s", student, module, assessment);

return get(student).flatMap(x -> x.get(module)).flatMap(x -> x.get(assessment)).map(x -> x.getGrade()).orElse(noRecord);
}

This is what I understand of the types in the chain of outputs:

Optional<Student> -> Optional<Module> -> Optional<Assessment> ...

does map(x -> x.getGrade()) return an Optional<String>? and if it return String only, how are we still able to use .orElse()?

**Comment by alantay11** [comment]
Afaik map will never return the actual object inside

**Comment by teresasee** [comment]
Hi, map(x -> x.getGrade()) takes the object inside the optional (which is the x) and maps x to x.getGrade(), which gives u Optional(grade). So you are still able to use .orElse().

**Comment by jlee1007** [comment]
Yep that will still return you an optional (as you can see from the test case).
It is orElse that will return a string! you can check javadocs for this too!

**Comment by Lu-Yi-Fan** [comment]
![Screenshot 2022-03-24 at 3 01 26 PM](https://user-images.githubusercontent.com/88626655/159860566-90c53ae7-59f5-4809-b8ce-ea7708f03433.png)

orElse is a method in Optional. So until .orElse() is evaluated, the object would be wrapped in an optional. orElse will return the value inside the optional or other (refer to screenshot)

**Comment by Earthstomp** [comment]
> Yep that will still return you an optional (as you can see from the test case). It is orElse that will return a string! you can check javadocs for this too!

![image](https://user-images.githubusercontent.com/81637338/159864685-8f33d8b5-3ef0-4e4c-9f2d-1900e0205f9a.png)

I think this might be what im confused with! Looking at codecrunch when I am able to getGrade() it will still return a string right? not Optional grade?

**Comment by Lu-Yi-Fan** [comment]
......map(x -> x.getGrade()) (Assessments getGrade()) return an Optional wrapped around the grade and the orElse will get the value out
